### PR TITLE
Theme loading modifications

### PR DESCRIPTION
**Pull request by sedge**
We force our way into the ThemeManager with hardcoded values for the Theme objects. We also had to override some CSS for the dark styles to prevent them from bleeding into the iframe.

### CODE REVIEW

File: src/extensions/default/bramble/main.js [hunk 1]
```diff
@@ -117,6 +123,13 @@ define(function (require, exports, module) {
             requestedPath = Path.resolve(currentDir, Path.normalize(message.path));
             XHRHandler.handleRequest(requestedPath);
         }
+
+        if(message.method === "ThemeToggle") {
+            var newTheme = currentTheme === 'light-theme' ? "dark-theme" : "light-theme";
+            currentTheme = newTheme;
+
+            PreferencesManager.set("themes.theme"
```
Comment (sedge) [inline src/extensions/default/bramble/main.js]: This should be `PreferencesManager.set("themes.theme", newTheme);`

File: src/utils/ExtensionLoader.js [hunk 2]
```diff
@@ -396,9 +397,17 @@ define(function (require, exports, module) {
         // Load *subset* of the usual builtin extensions list, and don't try to find any user/dev extensions
         if (brackets.inBrowser) {
             // Bramble: we have a more complex extension loading pattern based on URL params
-            var extensionList = BrambleExtensionLoader.getExtensionList(params);
+            var brambleExtensions = BrambleExtensionLoader.getExtensionList(params);
 
-            return Async.doInParallel(extensionList, function (item) {
+            return Async.doInParallel(brambleExtensions.extensionList, function (item) {
+                if (brambleExtensions.themes.indexOf(item) !== -1) {
```
Comment (sedge) [inline src/utils/ExtensionLoader.js]: This is almost correct, except `item` is an object. The string I need is on there.

File: src/extensions/default/bramble/main.js [hunk 3]
```diff
@@ -48,6 +50,10 @@ define(function (require, exports, module) {
     var defaultCSS  = require("text!lib/default-files/style.css");
     var defaultJS   = require("text!lib/default-files/script.txt");
 
+    // Store the current theme, and load defaults as if
```
Comment (sedge) [inline src/extensions/default/bramble/main.js]: This comment doesn't make sense anymore.

File: src/extensions/default/bramble/main.js [hunk 4]
```diff
@@ -22,6 +22,8 @@ define(function (require, exports, module) {
         UrlParams            = brackets.getModule("utils/UrlParams").UrlParams,
         Editor               = brackets.getModule("editor/Editor").Editor,
         Rewriter             = brackets.getModule("filesystem/impls/filer/lib/HTMLRewriter"),
+        ThemeManager         = brackets.getModule("view/ThemeManager"),
+        ThemePreferences     = PreferencesManager.getExtensionPrefs("themes"),
```
Comment (humphd) [inline src/extensions/default/bramble/main.js]: Make a new file in `lib/Theme.js` and require that in here.  Then, move all your theme pref flipping, etc into that.  This file is already way too big and messy.

File: src/utils/BrambleExtensionLoader.js [hunk 5]
```diff
@@ -9,9 +9,21 @@
...
+    var ThemeManager = brackets.getModule("/view/ThemeManager");
+
     var basePath = PathUtils.directory(window.location.href);
 
     /**
+     * Themes must be loaded seperately from the regular extension loading
+     * process because we load from a server vs the filesystem. Add new
+     * themes that should be loaded by default here.
+     */
+    var brambleDefaultThemes = [
```
Comment (humphd) [inline src/utils/BrambleExtensionLoader.js]: Maybe just `brambleThemes`, since we don't have anything other than our defaults (i.e., it's implied).

File: src/utils/BrambleExtensionLoader.js [hunk 6]
```diff
@@ -30,7 +42,7 @@ define(function (require, exports, module) {
         // Custom extensions we want loaded by default
         "bramble",
         "brackets-paste-and-indent"
-    ];
+    ].concat(brambleDefaultThemes);
```
Comment (humphd) [inline src/utils/BrambleExtensionLoader.js]: Don't do this here, do it in `processDefaults`.

File: src/utils/ExtensionLoader.js [hunk 7]
```diff
@@ -396,9 +397,17 @@ define(function (require, exports, module) {
...
+                    // XXXBramble: To make this more future proof,
+                    //  we need to read from package.json here to get
+                    //  a theme's metadata
```
Comment (humphd) [inline src/utils/ExtensionLoader.js]: You can read the package.json using `require(['text!' + ..., function(packageJSON){...}` inside `BrambleExtensionLoader`.  Maybe make `BrambleExtensionLoader.getExtensionList(params)` be async, and you can get those files before you return the list.

File: src/view/ThemeManager.js [hunk 8]
```diff
@@ -321,6 +322,44 @@ define(function (require, exports, module) {
         return loadFile(fileName, themePackage.metadata);
     }
 
+    /**
+     * Loads a remote theme by stubbing the File object expected by
```
Comment (humphd) [inline src/view/ThemeManager.js]: Add an `XXXBramble` to the start of this, so it's clear we did it, and why.

File: src/view/ThemeManager.js [hunk 9]
```diff
@@ -321,6 +322,44 @@ define(function (require, exports, module) {
...
+     *   theme: {
+     *     // Metadata here.
+     *   }
+     * }
+     *
+     * @param {Object} pre-populated file object.
+     * @param {Object} metadata options for the theme
+     * @return {Theme} the Theme object created for this theme
+     */
+     function manualAddTheme(file, options) {
```
Comment (humphd) [inline src/view/ThemeManager.js]: Just `addTheme()` is fine.  Everything you call in code is done so "manually", so it's not necessary.

**Comment by sedge** [comment]
Update: 

The overall architecture of the patch is complete, and I'm tracking down bugs as I test it. It's crashing on load at the moment.

File: src/view/ThemeManager.js [hunk 10]
```diff
@@ -358,13 +402,13 @@ define(function (require, exports, module) {
...
         refresh();
     });
-
     
     EventDispatcher.makeEventDispatcher(exports);
     
     exports.refresh         = refresh;
     exports.loadFile        = loadFile;
     exports.loadPackage     = loadPackage;
+    exports.addTheme        = addTheme;
```
Comment (JordanTheriault) [inline src/view/ThemeManager.js]: very small nit: Is there a reason why this is added here? It's strange that it wasn't appended to the end or alphabetically placed.

File: src/extensions/default/bramble/stylesheets/style.css [hunk 11]
```diff
@@ -136,6 +136,9 @@
 
 /* Iframe start */
 .iframeWidthHeight {
+    /* XXXBramble: Don't allow dark theme to bleed 
+         into the iframe */;
```
Comment (JordanTheriault) [inline src/extensions/default/bramble/stylesheets/style.css]: Can you condense this to one line?

File: src/extensions/default/bramble/lib/Theme.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
+/*jslint vars: true, plusplus: true, devel: true, nomen: true, indent: 4,
+maxerr: 50, browser: true */
+/*global define, brackets */
+
+/**
+ * This extension provides in-editor livepreview through an iframe,
+ * and leverages the experimental Multi Browser implementation of brackets
+ * (see https://github.com/adobe/brackets/tree/master/src/LiveDevelopment/MultiBrowserImpl)
```
Comment (humphd) [inline src/extensions/default/bramble/lib/Theme.js]: Um, your copy-paste is a bit too strong here.

File: src/extensions/default/bramble/lib/Theme.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    var Path             = brackets.getModule("filesystem/impls/filer/BracketsFiler").Path;
+    var basePath = PathUtils.directory(window.location.href);
+    var themePath = Path.join(basePath, 'extensions/default');
+
+    // Store the current theme, and load defaults as if
+    // they were third party themes
+    var currentTheme = 'light-theme';
+
+    function toggle(data) {
+        var newTheme;
```
Comment (humphd) [inline src/extensions/default/bramble/lib/Theme.js]: I don't understand the value of this, since you could just use `currentTheme` everywhere below and get the same result.  Remove it.

File: src/extensions/default/bramble/lib/Theme.js [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+        if (data.theme) {
+            currentTheme = newTheme = data.theme === 'light-theme' ? "dark-theme" : "light-theme";
+        } else {
+            currentTheme = newTheme = currentTheme === 'light-theme' ? "dark-theme" : "light-theme";
+        }
+
+        ThemePreferences.set("theme", newTheme);
```
Comment (humphd) [inline src/extensions/default/bramble/lib/Theme.js]: You're doing very similar things in this function to `setTheme`, so I'd recommend you call it instead.

File: src/extensions/default/bramble/lib/Theme.js [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+            theme: {
+                dark: true
+            }
+        };
+
+        ThemeManager.addTheme(lightFile, lightOptions);
+        ThemeManager.addTheme(darkFile, darkOptions);
+
+        currentTheme = currentTheme === 'light-theme' ? "dark-theme" : "light-theme";
+        ThemePreferences.set("theme", currentTheme);
```
Comment (humphd) [inline src/extensions/default/bramble/lib/Theme.js]: Use `setTheme` again here.

File: src/view/ThemeManager.js [hunk 16]
```diff
@@ -41,7 +41,7 @@ define(function (require, exports, module) {
     var loadedThemes    = {},
         currentTheme    = null,
         styleNode       = $(ExtensionUtils.addEmbeddedStyleSheet("")),
-        defaultTheme    = "thor-light-theme",
```
Comment (humphd) [inline src/view/ThemeManager.js]: What is this theme?  Brackets doesn't use `Light Theme` by default?

File: src/view/ThemeManager.js [hunk 17]
```diff
@@ -321,6 +324,49 @@ define(function (require, exports, module) {
         return loadFile(fileName, themePackage.metadata);
     }
 
+    /**
+     * XXXBramble: Required for loading themes from the server vs.
+     *             the filesystem (the brackets default)
```
Comment (humphd) [inline src/view/ThemeManager.js]: Remove this indentation.

File: src/view/ThemeManager.js (hunk 16 above)
Comment (sedge) [inline src/view/ThemeManager.js]: Apparently not. That's the only reference to it in the `/src` directory.

File: src/view/ThemeManager.js [hunk 18]
```diff
@@ -321,6 +324,49 @@ define(function (require, exports, module) {
...
+     *   name: {String} filename of the theme's LESS file
+     *   _path: {String} URL path to the theme's LESS file
+     * }
+     *
+     * Likewise, the options object must contain whatever metadata
+     * is found in the theme's package.json file:
+     *
+     * {
+     *   theme: {
+     *     // Metadata here.
```
Comment (humphd) [inline src/view/ThemeManager.js]: Mention `dark` here, at least.

File: src/view/ThemeManager.js [hunk 19]
```diff
@@ -358,13 +404,13 @@ define(function (require, exports, module) {
     EditorManager.on("activeEditorChange", function () {
         refresh();
     });
-
```
Comment (humphd) [inline src/view/ThemeManager.js]: Undo this change

File: src/extensions/default/bramble/lib/Theme.js (hunk 13 above)
Comment (humphd) [inline src/extensions/default/bramble/lib/Theme.js]: I think you can do this in 2 lines:

```js
var theme = data.theme || currentTheme;
setTheme(theme === "light-theme" ? "dark-theme" : "light-theme");
```

**Comment by humphd** [comment]
r+ with travis plus confirmation it all still works

File: src/extensions/default/bramble/stylesheets/style.css [hunk 20]
```diff
@@ -136,6 +136,8 @@
 
 /* Iframe start */
 .iframeWidthHeight {
+    /* XXXBramble: Don't allow dark theme to bleed into the iframe */;
```
Comment (JordanTheriault) [inline src/extensions/default/bramble/stylesheets/style.css]: // vs /* */ for 1 line

Comment (sedge) [inline src/extensions/default/bramble/stylesheets/style.css]: Doesn't work in CSS

**Comment by JordanTheriault** [comment]
r+, great job!
